### PR TITLE
ISPN-14594 SADD plus internal impl for set

### PR DESCRIPTION
**Pull request by rigazilla**
https://issues.redhat.com/browse/ISPN-14594

initial dev for set commands

### CODE REVIEW

File: multimap/src/main/java/org/infinispan/multimap/impl/EmbeddedSetCache.java [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+import org.infinispan.container.impl.InternalEntryFactory;
+import org.infinispan.functional.FunctionalMap;
+import org.infinispan.functional.impl.FunctionalMapImpl;
+import org.infinispan.functional.impl.ReadWriteMapImpl;
+import org.infinispan.multimap.impl.function.AddFunction;
+
+/**
+ * Multimap with Linked List Implementation methods
+ *
+ * @author  Katia Aresti
```
Comment (karesti) [inline multimap/src/main/java/org/infinispan/multimap/impl/EmbeddedSetCache.java:21]: Vittoriooooooooo

File: multimap/src/main/java/org/infinispan/multimap/impl/SetBucket.java [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+   @ProtoFactory
+   SetBucket(Collection<MarshallableUserObject<V>> wrappedValues) {
+      this((Set<V>) wrappedValues.stream().map(MarshallableUserObject::get)
+            .collect(Collectors.toCollection(HashSet::new)));
+   }
+
+   @ProtoField(number = 1, collectionImplementation = ArrayList.class)
+   Collection<MarshallableUserObject<V>> getWrappedValues() {
+      return this.values.stream().map(MarshallableUserObject::new).collect(Collectors.toCollection(ArrayDeque::new));
```
Comment (karesti) [inline multimap/src/main/java/org/infinispan/multimap/impl/SetBucket.java:55]: not needed. is a set. you need to check the Bucket implementation

File: multimap/src/main/java/org/infinispan/multimap/impl/function/AddFunction.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+   private static class Externalizer implements AdvancedExternalizer<AddFunction> {
+
+      @Override
+      public Set<Class<? extends AddFunction>> getTypeClasses() {
+         return Collections.singleton(AddFunction.class);
+      }
+
+      @Override
+      public Integer getId() {
+         return ExternalizerIds.OFFER_FUNCTION;
```
Comment (karesti) [inline multimap/src/main/java/org/infinispan/multimap/impl/function/AddFunction.java:60]: id needs to change

File: server/resp/src/main/java/org/infinispan/server/resp/commands/set/SADD.java [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+
+   protected CompletionStage<RespRequestHandler> addAndReturn(Resp3Handler handler,
+                                                               ChannelHandlerContext ctx,
+                                                               List<byte[]> arguments) {
+      byte[] key = arguments.get(0);
+      EmbeddedSetCache<byte[], byte[]> esc = handler.getEmbeddedSetCache();
+      AggregateCompletionStage<Void> aggregateCompletionStage = CompletionStages.aggregateCompletionStage();
+      var addedCount = new AtomicLong(0);
+      for (int i = 1; i < arguments.size(); i++) {
+         CompletionStage<Void> push = esc.add(key, arguments.get(i)).thenAccept((v) -> {if (v) addedCount.incrementAndGet();});
```
Comment (karesti) [inline server/resp/src/main/java/org/infinispan/server/resp/commands/set/SADD.java:52]: rename to add

**Review by karesti — CHANGES_REQUESTED** [review]
you need to add some test for distributed mode as well, to execute tests from owner and not owner

File: multimap/src/main/java/org/infinispan/multimap/impl/SetBucket.java (hunk 2 above)
Comment (rigazilla) [reply multimap/src/main/java/org/infinispan/multimap/impl/SetBucket.java:55]: you mean change it to:
```
      return this.values.stream().map(MarshallableUserObject::new).collect(Collectors.toSet());
```
?

File: server/resp/src/test/java/org/infinispan/server/resp/RespSetCommandsTest.java [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+
+import org.testng.annotations.Test;
+
+import io.lettuce.core.api.sync.RedisCommands;
+
+@Test(groups = "functional", testName = "server.resp.RespSetCommandsTest")
+public class RespSetCommandsTest extends SingleNodeRespBaseTest {
+    @Test
+    public void testSadd() {
+        RedisCommands<String, String> redis = redisConnection.sync();
```
Comment (karesti) [inline server/resp/src/test/java/org/infinispan/server/resp/RespSetCommandsTest.java:13]: we need more tests covering (check lists tests) 
if the key already exist and not a set, should throw an error

**Review by karesti — CHANGES_REQUESTED** [review]

**Comment by rigazilla** [comment]
> you need to add some test for distributed mode as well, to execute tests from owner and not owner

there's any example I can use as reference?

**Comment by karesti** [comment]
DistributedMultimapListCacheTest @rigazilla

**Comment by rigazilla** [comment]
> DistributedMultimapListCacheTest @rigazilla

added DistributedSetTest.java

**Review by karesti — DISMISSED** [review]
LGTM but needs rebase

**Pull request merged**